### PR TITLE
feat: add safer Dec struct for math operations

### PR DESCRIPTION
**Pull request by aaronc**
In https://github.com/regen-network/mainnet/pull/15, it was discovered that using `apd.Decimal` directly can be unsafe because apd operations mutate the underlying `Decimal`, but when copying the `big.Int` structure can be shared between `Decimal` instances causing corruption. This PR introduces a `Dec` wrapper which does NO mutation of `apd.Decimal`s, instead creating a new `apd.Decimal` for every operation so that usage is safe.

### CODE REVIEW

**Comment by clevinson** [comment]
Marking this PR as ready for review, after adding a few additional test and updating to master.

Next follow-ups after this afaiu are:
- add property based test for new `Dec` struct
- refactor all usages of `ParseNonNegativeFixedDecimal`, etc. to be based off of the internal `Dec` struct instead (I can do this, but might be worth waiting until #428 is merged as this changes some of the precision functionality)

File: types/math/dec.go
Comment (robert-zaremba) [inline types/math/dec.go:13]: Why 34. Can we add a comment to it (eg link to a related discussion)?

File: types/math/dec.go
Comment (robert-zaremba) [inline types/math/dec.go:46]:
```suggestion
// SafeSub subtracts the value of y from x and stores the result in res with arbitrary precision.
// Returns ErrInsufficientFunds error if the result is negative.
```

File: types/math/dec.go
Comment (robert-zaremba) [inline types/math/dec.go:33]:
```suggestion
// Add returns new Dec with value `x+y` without mutating any argument and error if 
// there is an overflow.
func (x Dec) Add(y Dec) (Dec, error) {
```

Comment (robert-zaremba) [reply types/math/dec.go:33]: Shall we add similar comment to other functions?

File: types/math/dec.go
Comment (robert-zaremba) [inline types/math/dec.go:35]: let's add `SafeAdd` or check here (in `Add`) that both `x` and `y` are not negative.

**Review by robert-zaremba — COMMENTED** [review]
Looks good. Left few cosmetic comment and one concern about `Add` (maybe we should have `SafeAdd` as well).

File: types/math/dec_test.go
Comment (ruhatch) [inline types/math/dec_test.go:9]: Yup, this looks like a good spot for some property tests :D

Even if we don't add those in this PR, perhaps it would be good to identify some properties we'd expect these types to have? They can just be in comments in some basic mathematical notation for now.

**Review by ruhatch — CHANGES_REQUESTED** [review]
I'd like to see some more documentation about this new type and why we're using it, plus some basic outlines of the properties we expect to hold

File: types/math/dec.go
Comment (clevinson) [reply types/math/dec.go:35]: What's wrong with either `x` or `y` being negative? There's tests in here that include usage of negative `Dec` values, I'm not sure why we would want to enforce at the library level that this function cannot work with negative values.

Can you explain the use case for a `SafeAdd` ? `SafeSub` makes more sense to me, as verifying that balance is non-zero is in fact something we often have to do. I can't really think of a use case for `SafeAdd` though.

File: types/math/dec.go
Comment (clevinson) [reply types/math/dec.go:8]: Why do you say we're trying to use them as unsigned? I think we only are restricting that they're unsigned in `SafeSub`, which will be used sometimes when dealing with token balances (and wanting to ensure a non-negative result when subtracting two numbers).  But more generally speaking, I don't see why we would want to restrict this struct to only work with unsigned decimals.  There could be valid reasons for us storing ecological measurements as negative numbers (e.g. temperature) at some point down the line.

File: types/math/dec.go
Comment (clevinson) [reply types/math/dec.go:46]: This is done!

File: types/math/dec.go
Comment (clevinson) [reply types/math/dec.go:24]: Done!

File: types/math/dec.go
Comment (clevinson) [reply types/math/dec.go:15]: Just updated with documentation about the wrapper (quoting from @aaronc in the original PR description).

File: types/math/dec_test.go
Comment (clevinson) [reply types/math/dec_test.go:9]: @ruhatch  Just added some suggestions as comments in the bottom of `TestDec()`. I'm pretty new to property tests so let me know what you think?

If the decimal nature of this arithmetic means that actual equality isn't really feasible to test, then perhaps we can property test to keep error range as minimal as possible? e.g. `(x/y)*y - x) < 1e-15` or something similar?

File: types/math/dec.go
Comment (ruhatch) [reply types/math/dec.go:15]: > Why do you say we're trying to use them as unsigned? I think we only are restricting that they're unsigned in `SafeSub`, which will be used sometimes when dealing with token balances (and wanting to ensure a non-negative result when subtracting two numbers). But more generally speaking, I don't see why we would want to restrict this struct to only work with unsigned decimals. There could be valid reasons for us storing ecological measurements as negative numbers (e.g. temperature) at some point down the line.

Okay, well then I think we're doing some mixing of concepts here. I don't think the decimal type should be holding information about the fact that it will be used as a non-negative token balance. I.e. I don't think an `InsufficientFunds` error should be thrown from within the API of this type. I also then don't think the name `SafeSub` is quite accurate, because it's not unsafe to subtract and get a negative result.

I guess there's a few options here:
- Have a separate type for explicitly unsigned decimals, for cases such as token balances where they'll never be negative
- Rename `SafeSub` to something like `SubNonNegative` (not catchy)
- Don't expose a `SafeSub` function and do the checking and error throwing elsewhere

Does that make sense?

Comment (ruhatch) [reply types/math/dec.go:15]: Personally I'm in favour of the first option. That way we're explicit in the type system about the kind of values that we're representing.

File: types/math/dec.go
Comment (ruhatch) [reply types/math/dec.go:48]: I think this relates to https://github.com/regen-network/regen-ledger/pull/307/files#r678209627

File: types/math/dec_test.go
Comment (ruhatch) [inline types/math/dec_test.go:90]: These looks nice, thanks :)

**Review by ruhatch — COMMENTED** [review]
Looks good, just need to make a decision about the `SafeSub` stuff

File: types/math/dec_test.go
Comment (ruhatch) [reply types/math/dec_test.go:9]: Yeah, some error range property would be great. I'm not sure what the standard should be, as I'm not super experienced with float stuff!

File: types/math/dec.go
Comment (ruhatch) [reply types/math/dec.go:15]: I think we could also then have a type alias for token balances that returns the `ErrInsufficientFunds` error in it's own `SafeSub` method.

```
type Balance = UDec

func (*Balance) SafeSub ...
```

File: types/math/dec.go
Comment (ruhatch) [inline types/math/dec.go:84]:
```suggestion
// QuoInteger returns a new integral Dec with value `x/y` (formatted as decimal128, with 34 digit precision)
```

File: types/math/dec.go
Comment (ruhatch) [inline types/math/dec.go:92]:
```suggestion
// Rem returns the integral remainder from `x/y` (formatted as decimal128, with 34 digit precision) without
```

File: types/math/dec.go
Comment (clevinson) [reply types/math/dec.go:15]: So I opted for moving this just out of Dec struct, and just migrated to a function in the math package with signature:
```go
func SubBalance(x Dec, y Dec) (Dec, error)
```

Seems like @aaronc also had desire for a SafeAdd type operation, so I've also implemented an `AddBalance()` similar to the `SubBalance()` above.

I do think we should probably think about refactoring these, but probably better to do when we're actually looking at how these functions are used in context (presumably when actually performing arithmetic on ecocredits). So I'd like to merge this in, and when we get to making use of the `Dec` struct for tracking ecocredit balances, we then take another look at where the "safe" arithmetic options should be.

Also holding that from @aaronc's comments today, it seems that we're trying to align the implementation here with a yet-to-be completed refactor of `sdk.Dec` (note: sdk.Dec does not have an implementation of `SafeSub()`, but [sdk.DecCoin does](https://github.com/cosmos/cosmos-sdk/blob/master/types/dec_coin.go#L312-L317)).  I don't think it makes sense to have a `DecCoin` struct in this package (as ecocredits are not necessarily coins...), but maybe it makes sense to have some struct with more explicit coin-like behavior (e.g. treating it more like an unsigned dec) when refactoring the actual balance tracking to use the Dec?

Comment (ruhatch) [reply types/math/dec.go:15]: Yeah, I think at some point an explicitly unsigned Dec type would be good. Then we can property test it properly and know that we have some nice safety guarantees on it.

File: types/math/math.go
Comment (ruhatch) [inline types/math/math.go:139]: Let's do this before doing the actual addition to avoid an unnecessary operation!

**Review by ruhatch — APPROVED** [review]
Just one small thing in `AddBalance`, but preapproving

File: types/math/math.go
Comment (aaronc) [inline types/math/math.go:116]:
```suggestion
func SafeSubBalance(x Dec, y Dec) (Dec, error) {
```

File: types/math/math.go
Comment (aaronc) [inline types/math/math.go:132]:
```suggestion
func SafeAddBalance(x Dec, y Dec) (Dec, error) {
```

**Review by aaronc — COMMENTED** [review]
Can't approve for some reason on mobile but ACK

File: types/math/dec.go
Comment (robert-zaremba) [reply types/math/dec.go:15]: I prefer `Safe*` - this is what we have in other blockchain and it's more familiar to developers.

File: types/math/dec.go
Comment (robert-zaremba) [reply types/math/dec.go:48]: > What's wrong with either x or y being negative?

@clevinson  we want to have "Safe" operations - operations which are assured to not return negative values. `2 + (-3)` is negative.

File: types/math/math.go
Comment (robert-zaremba) [reply types/math/math.go:116]: Do we need `Balance` suffix? It doesn't need to be related to a balance. I think `SafeSub` or `CheckedSub` is good enough.

**Review by robert-zaremba — COMMENTED** [review]
There is a proposal to rename `SubBalance` to `SafeSubBalance`. I think it's a bit of over-engineering. The operation doesn't need to be related to balance. It could be a weight for example (it's not possible to have a negative weight). `Safe*` is widely adopted in Ethereum dev community. In Rust they use `check*`

**Comment by clevinson** [comment]
Keeping with `SafeSubBalance` for now as the error is specific to coins (insufficient funds error), and `SafeAddBalance` should use the same language IMO as `SafeSubBalance`.

I'm open to refactoring once we have a more clear context in which these functions are used, but it seems difficult to bike shed over this API naming when we're not actually seeing its usage (e.g. maybe a different package besides math.go is where this belongs?).  I also like the idea from @ruhatch about tying these functions to unsigned math types.

**Review by clevinson — APPROVED** [review]
Approving in place of @aaronc's ACK (@aaronc doesn't have approve privileges, as he was original author)

**Pull request merged**
